content/blog/updates/2024-01-09-incident/index.md

```diff
@@ -0,0 +1,84 @@
+
+For historical reasons, most requests that Budibase makes to its internal database first check to make sure that the database they're writing to exists. If it does not, it gets created. For the 6 minutes that `chesterfield-004` was in the load balancer, most requests to check that a database existed would have reported that no, the database does not exist. A new, empty database would be created and the code would continue. This empty database was then replicated to other nodes, effectively deleting all of their contents. We were able to reproduce this behaviour locally, and you can check our methodology [here][4].
+
+{{< figure src="chesterfield-004-writes.png" alt="A graph showing the number of writes per second to chesterfield-004. It peaks at 1 write per second at 13:54, then trails off until the line disappears at 13:58">}}
+
+This manifested for most users as the inability to log in, as their user record had been deleted.
+
+{{< figure src="login-403s.png" alt="A graph showing HTTP requests to Budibase Cloud that had a 403 response, indicating login failure" >}}
+
+Unfortunately we were not aware that there was a problem until 14:13 UTC, when we started receiving reports from customers that they were not able to log in. Why, then, had we removed the node from the cluster and turned it down?
+
+When our automation brought up the node, it failed at the point where it was installing CouchDB. We investigated this at the time and found it had used an incorrect CouchDB version string when installing the CouchDB package. We attempted to correct this and re-run the automation, but that didn't work. This prompted us to bring the node down completely, and start again from the beginning.
+
+This was both lucky and unfortunate. Lucky because we inadvertently fixed what was causing data corruption. The data that had been lost was still gone, but we weren't losing more data. Unfortunate because when this node went, it took valuable debugging information with it.
+
+# How we responded
+
+The first customer reports got to us in the engineering team at 14:13 UTC, and we immediately suspected our work on the database cluster was the cause. The timing lined up perfectly, and it was a substantial change to our infrastructure. By 14:28 UTC we had confirmed data loss in the tenants that had reached out to us and began executing our disaster recovery plan.
+
+We managed to confirm data loss in 11 customers by looking at our records of failed logins, and given that the rate of failed logins was not increasing we believed that without `chesterfield-004` in the cluster, the problem was not spreading. However, 11 tenants was a lower bound. It was difficult to know what writes had been served by `chesterfield-004` prior to it being taken down, and the log files on the machine had not been preserved. Because of this, we decided to play it safe and restore all data to the last known-good backup we had, which was from 13:33 UTC.
```

Comment:
Thankfully due to the way the DBs were accessed, we were able to recover essentially all of the corrupted DBs. The problem that generally happened with "missing apps" was caused when trying to access a certain DB (usually the production app) and this failed, causing the cache to be populated with an invalid JSON. After us clearing the relevant keys in the redis cache manually, users were able to access their apps again.

I'm not sure the same can be said for rows created during that window of course.

Reply:
I'm not sure exactly what you mean here. We recovered the corrupt DBs with the EBS snapshot restore, then there was still a problem after the nodes came back up? I can't remember that.

Reply:
Yep, some users reported some of their apps were missing. This is due to the way we store `app_metadata` documents. We cache them in redis so we don't have to go fetching it again as it rarely changes.

During the outage, any user who tried to access one of their apps and got routed to `chesterfield-004` had their `app_metadata` redis keys populated with a JSON something like this `{  status: "invalid" }` so when trying to load the apps, any of those `corrupted` apps were missing.

By deleting the invalid keys in redis after the snapshot restore, they could access their apps again as they were simply refetched from the DB.

Reply:
Have mentioned this near the end of the post.